### PR TITLE
8221785: Let possibly_parallel_threads_do cover the same threads as threads_do

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that makes `Threads::possibly_parallel_threads_do` iterate over the same set of threads as `threads_do` to have parity? I.e. over all java and non-java threads.

Originally this CR has been created to make a new method that keeps iterating only over java threads and the VM thread, but it's a bit weird to have both variants as the overhead of the extra threads is negligible and otherwise just confusing.

So I made `Threads::possibly_parallel_threads_do` iterate over all threads; all uses support that afaict, also all uses correctly change the claim token (mostly in the enclosing `StrongRootsScope`).

This allows some minimally better hiding of the token mechanism.

One other reason for not doing this in the first place (as in [JDK-8221102](https://bugs.openjdk.org/browse/JDK-8221102), or as discussed [here](https://mail.openjdk.org/pipermail/hotspot-dev/2019-April/037541.html)) has been the fear that there would be a problem with threads being created during iteration and the (common) call to 'Threads::assert_all_threads_claimed`. However all calls so far are during a safepoint, and none seem to create new threads. I suggest to defer looking at this problem when it is important.

Moreover I need that functionality is required for (JDK-8211104)[https://bugs.openjdk.org/browse/JDK-8211104]. :)

Testing: tier1-4, gha

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8221785](https://bugs.openjdk.org/browse/JDK-8221785): Let possibly_parallel_threads_do cover the same threads as threads_do

### Reviewers
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [34e67f56](https://git.openjdk.org/jdk/pull/12201/files/34e67f56a1de6136c0e1dad4fd3ad97bc0895dcc)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [34e67f56](https://git.openjdk.org/jdk/pull/12201/files/34e67f56a1de6136c0e1dad4fd3ad97bc0895dcc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12201/head:pull/12201` \
`$ git checkout pull/12201`

Update a local copy of the PR: \
`$ git checkout pull/12201` \
`$ git pull https://git.openjdk.org/jdk pull/12201/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12201`

View PR using the GUI difftool: \
`$ git pr show -t 12201`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12201.diff">https://git.openjdk.org/jdk/pull/12201.diff</a>

</details>

### CODE REVIEW

**Review by walulyai — APPROVED** [review]
Lgtm!

File: src/hotspot/share/runtime/threads.cpp [hunk 1]
```diff
@@ -260,9 +260,11 @@ void Threads::possibly_parallel_threads_do(bool is_par, ThreadClosure* tc) {
...
     }
   }
-  VMThread* vmt = VMThread::vm_thread();
-  if (vmt->claim_threads_do(is_par, claim_token)) {
-    tc->do_thread(vmt);
+  for (NonJavaThread::Iterator njti; !njti.end(); njti.step()) {
+    Thread* current = njti.current();
+    if (current->claim_threads_do(is_par, claim_token)) {
+      tc->do_thread(current);
+    }
```
Comment (coleenp) [inline src/hotspot/share/runtime/threads.cpp:267]: If I understand correctly, if this is only OK in a safepoint, can you add the safepoint assert here?

**Review by coleenp — APPROVED** [review]
Thank you for picking this up.

**Comment by tschatzl** [comment]
Thanks @walulyai @coleenp for your reviews
/integrate

**Pull request closed without merging**